Non-ssl Postgres connection fails

Hey folks,

I'm trying to run Metabase against a Postgres installation that does not need/support SSL (because all traffic can be trusted.)  The docs say Metabase will prefer SSL but fall back to a regular connection, but that doesn't seem to be happening.

![](http://screens.dashkb.net/20160208183239.png)

I grepped around for `ssl` and set random things to false, shooting in the dark, rebuilding the jar, but no luck.

There is no better info in the Metabase logs, btw.

```
2016-02-09T17:59:12.277894Z app[web.ip17231230-42856c56-a55d-4c89-8351-0a1ebd4ca2ca]: 2016-02-09 17:59:12,277 [DEBUG] metabase.db.internal :: DB CALL:  User [:id :email :date_joined :first_name :last_name :last_login :is_superuser :is_active :is_staff] (where (limit (select* ENTITY23107) 1) {:id current-user-id})
2016-02-09T17:59:12.283219Z app[web.ip17231230-42856c56-a55d-4c89-8351-0a1ebd4ca2ca]: 2016-02-09 17:59:12,283 [ERROR] metabase.driver :: Failed to connect to database: org.postgresql.util.PSQLException: The server does not support SSL.
2016-02-09T17:59:12.286883Z app[web.ip17231230-42856c56-a55d-4c89-8351-0a1ebd4ca2ca]: 2016-02-09 17:59:12,286 [ERROR] metabase.driver :: Failed to connect to database: org.postgresql.util.PSQLException: The server does not support SSL.
2016-02-09T17:59:12.302218Z app[web.ip17231230-42856c56-a55d-4c89-8351-0a1ebd4ca2ca]: 2016-02-09 17:59:12,302 [DEBUG] metabase.middleware :: POST /api/database 400 (10 ms)
2016-02-09T17:59:12.302252Z app[web.ip17231230-42856c56-a55d-4c89-8351-0a1ebd4ca2ca]: {:valid false,
2016-02-09T17:59:12.302423Z app[web.ip17231230-42856c56-a55d-4c89-8351-0a1ebd4ca2ca]:  :dbname
2016-02-09T17:59:12.302578Z app[web.ip17231230-42856c56-a55d-4c89-8351-0a1ebd4ca2ca]:  "org.postgresql.util.PSQLException: The server does not support SSL.",
2016-02-09T17:59:12.302604Z app[web.ip17231230-42856c56-a55d-4c89-8351-0a1ebd4ca2ca]:  :message
2016-02-09T17:59:12.302719Z app[web.ip17231230-42856c56-a55d-4c89-8351-0a1ebd4ca2ca]:  "org.postgresql.util.PSQLException: The server does not support SSL."}
2016-02-09T17:59:12.302874Z app[web.ip17231230-42856c56-a55d-4c89-8351-0a1ebd4ca2ca]:
```

The Flynn cluster's Postgres log is showing connections, but no authentication. Here is the log from me hitting save in Metabase once:

```
2016-02-09T22:13:31.887606Z app[postgres.flynn14548659832-f8679bd2-0f4f-475a-9c0d-4046f01ddf2f]: ad8b4338-b271-4953-b921-59809ef13fa3 2016-02-09 22:13:31.887 UTC LOG:  connection received: host=100.100.9.0 port=59670
2016-02-09T22:13:31.891244Z app[postgres.flynn14548659832-f8679bd2-0f4f-475a-9c0d-4046f01ddf2f]: ad8b4338-b271-4953-b921-59809ef13fa3 2016-02-09 22:13:31.891 UTC LOG:  connection received: host=100.100.9.0 port=59671
```

I had assumed that the first connection is the failed SSL connection, and the second is non-SSL. What is strange is that the Metabase logs for that show the same message twice as well:

```
2016-02-09T22:14:46.646721Z app[web.flynn14548659831-c3af81ae-3c57-4233-8111-97331416e0c2]: 2016-02-09 22:14:46,645 [DEBUG] metabase.db.internal :: DB CALL:  User [:id :email :date_joined :first_name :last_name :last_login :is_superuser :is_active :is_staff] (where (limit (select* ENTITY23107) 1) {:id current-user-id})
2016-02-09T22:14:46.683646Z app[web.flynn14548659831-c3af81ae-3c57-4233-8111-97331416e0c2]: 2016-02-09 22:14:46,683 [ERROR] metabase.driver :: Failed to connect to database: org.postgresql.util.PSQLException: The server does not support SSL.
2016-02-09T22:14:46.695594Z app[web.flynn14548659831-c3af81ae-3c57-4233-8111-97331416e0c2]: 2016-02-09 22:14:46,695 [ERROR] metabase.driver :: Failed to connect to database: org.postgresql.util.PSQLException: The server does not support SSL.
2016-02-09T22:14:46.719150Z app[web.flynn14548659831-c3af81ae-3c57-4233-8111-97331416e0c2]: 2016-02-09 22:14:46,718 [DEBUG] metabase.middleware :: POST /api/database 400 (62 ms)
2016-02-09T22:14:46.719281Z app[web.flynn14548659831-c3af81ae-3c57-4233-8111-97331416e0c2]: {:valid false,
2016-02-09T22:14:46.719643Z app[web.flynn14548659831-c3af81ae-3c57-4233-8111-97331416e0c2]:  :dbname
2016-02-09T22:14:46.719937Z app[web.flynn14548659831-c3af81ae-3c57-4233-8111-97331416e0c2]:  "org.postgresql.util.PSQLException: The server does not support SSL.",
2016-02-09T22:14:46.720228Z app[web.flynn14548659831-c3af81ae-3c57-4233-8111-97331416e0c2]:  :message
2016-02-09T22:14:46.720419Z app[web.flynn14548659831-c3af81ae-3c57-4233-8111-97331416e0c2]:  "org.postgresql.util.PSQLException: The server does not support SSL."}
2016-02-09T22:14:46.720635Z app[web.flynn14548659831-c3af81ae-3c57-4233-8111-97331416e0c2]:
```

Unfortunately that is all we have to go on right now as the logs don't give us more information. 

Does metabase require any specific user permissions? Flynn provisions a unique set of credentials for each app, so it does not have full (e.g. admin) permissions.

FWIW, the same credentials work fine with [pgweb](https://github.com/sosedoff/pgweb) in Flynn, with SSL explicitly disabled in the connection dialog.

--- a/src/metabase/driver/postgres.clj
+++ b/src/metabase/driver/postgres.clj
@@ -87,17 +87,21 @@
    :sslmode    "require"
    :sslfactory "org.postgresql.ssl.NonValidatingFactory"})  ; HACK Why enable SSL if we disable certificate validation?
 
+(def ^:const disable-ssl-params
+  "Params to include in the JDBC connection spec to disable SSL."
+  {:sslmode "disable"})
+
 (defn- connection-details->spec [_ {:keys [ssl] :as details-map}]
   (-> details-map
       (update :port (fn [port]
                       (if (string? port) (Integer/parseInt port)
                           port)))
       (dissoc :ssl)               ; remove :ssl in case it's false; DB will still try (& fail) to connect if the key is there
-      (merge (when ssl            ; merging ssl-params will add :ssl back in if desirable
-               ssl-params))
+      (merge (if ssl
+               ssl-params
+               disable-ssl-params))
       (rename-keys {:dbname :db})
       kdb/postgres))
-
 
 (defn- unix-timestamp->timestamp [_ expr seconds-or-milliseconds]
   (case seconds-or-milliseconds

--- a/test/metabase/driver/postgres_test.clj
+++ b/test/metabase/driver/postgres_test.clj
@@ -15,7 +15,8 @@
    :classname   "org.postgresql.Driver"
    :subprotocol "postgresql"
    :subname     "//localhost:5432/bird_sightings"
-   :make-pool?  true}
+   :make-pool?  true
+   :sslmode     "disable"}
   (sql/connection-details->spec (PostgresDriver.) {:ssl    false
                                                    :host   "localhost"
                                                    :port   5432

